src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialHelper.cs

```diff
@@ -0,0 +1,112 @@
+        /// </summary>
+        /// <returns>Returns true if the previously cached client assertion is valid</returns>
+        public static bool ValidateClientAssertion(ClientCredential clientCredential, AuthorityEndpoints endpoints, bool sendX5C)
+        {
+            if (clientCredential == null)
+            {
+                throw new ArgumentNullException(nameof(clientCredential));
+            }
+            else if (string.IsNullOrWhiteSpace(clientCredential.Assertion))
+            {
+                return false;
+            }
+
+            //Check if all current client assertion values match incoming parameters and expiration time
+            //The clientCredential object contains the previously used values in the cached client assertion string
+            bool expired = clientCredential.ValidTo <=
+                           JsonWebToken.ConvertToTimeT(
+                               DateTime.UtcNow + TimeSpan.FromMinutes(Constants.ExpirationMarginInMinutes));
+
+            bool parametersMatch = clientCredential.Audience == endpoints?.SelfSignedJwtAudience &&
+                                   clientCredential.ContainsX5C == sendX5C;
+
+            return !expired && parametersMatch;
+        }
+
+        public static Dictionary<string, string> CreateClientCredentialBodyParameters(
+            ICoreLogger logger, 
+            ICryptographyManager cryptographyManager, 
+            ClientCredential clientCredential,
+            string clientId,
+            AuthorityEndpoints endpoints, 
+            bool sendX5C)
+        {
+            Dictionary<string, string> parameters = new Dictionary<string, string>();
+            if (clientCredential != null)
+            {
+                if (!string.IsNullOrEmpty(clientCredential.Secret))
+                {
+                    parameters[OAuth2Parameter.ClientSecret] = clientCredential.Secret;
+                }
+                else
+                {
+                    if (clientCredential.Assertion == null || clientCredential.ValidTo != 0)
+                    {
+                        if (!ValidateClientAssertion(clientCredential, endpoints, sendX5C))
+                        {
+                            logger.Info("Client Assertion does not exist or near expiry.");
+                            var jwtToken = new JsonWebToken(cryptographyManager, clientId, endpoints?.SelfSignedJwtAudience);
+                            clientCredential.Assertion = jwtToken.Sign(clientCredential.Certificate, sendX5C);
+                            clientCredential.ValidTo = jwtToken.Payload.ValidTo;
+                            clientCredential.ContainsX5C = sendX5C;
+                            clientCredential.Audience = endpoints?.SelfSignedJwtAudience;
+                        }
+                        else
+                        {
+                            logger.Info("Reusing the unexpired Client Assertion...");
```

Comment:
>"Reusing the unexpired Client Assertion..." [](start = 40, length = 43)

logMessage.cs (a few places)... though we will have a discussion remaining on this.

Reply:
This is an existing logger message so I haven't been migrating these as part of this PR since it's already large.  We should do the logger cleanup (after our discussion) separately.  

---
In reply to: [249090557](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/798#discussion_r249090557) [](ancestors = 249090557)